### PR TITLE
#105 memory leaks

### PR DESCRIPTION
**Pull request by Alex009**

### CODE REVIEW

File: buildSrc/src/main/kotlin/Deps.kt [hunk 1]
```diff
@@ -1,21 +1,22 @@
 object Deps {
-    private const val kotlinVersion = "1.4.21"
+    private const val kotlinVersion = "1.4.30"
```
Comment (Alex009) [inline buildSrc/src/main/kotlin/Deps.kt:2]: here should be 31 now. also rename to more concrete usage (kotlin version controlled by build.gradle in buildSrc now)

File: mvvm-livedata/src/androidMain/kotlin/dev/icerock/moko/mvvm/livedata/LiveData.kt [hunk 2]
```diff
@@ -4,35 +4,74 @@
 
 package dev.icerock.moko.mvvm.livedata
 
-import androidx.lifecycle.Observer
-import androidx.lifecycle.LiveData as ArchLiveData
+// TODO commonize logic of Android and iOS without broking of API and ABI.
```
Comment (Alex009) [inline mvvm-livedata/src/androidMain/kotlin/dev/icerock/moko/mvvm/livedata/LiveData.kt:7]: todo now or later?

File: mvvm-livedata/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/FlatMapTest.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        val input2: MutableLiveData<Int> = MutableLiveData(initialValue = 1)
+        val output: LiveData<String> = input.flatMapBuffered { current, new ->
+            MutableLiveData("$current:$new").mergeWith(input2) { str, int ->
+                "$str:$int"
+            }
+        }
+        val observer = AssertObserver<String>()
+        output.addObserver(observer)
+
+        // TODO maybe we should change logic of flatMapBuffered at initialization step?
```
Comment (Alex009) [inline mvvm-livedata/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/FlatMapTest.kt:92]: todo now or later?

File: mvvm-livedata/src/iosMain/cinterop/testproj/ViewController.m [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+@end
+
+@implementation LifecycleDelegate
+
+- (void)view:(UIView *)view willMoveToWindow:(UIWindow *)window {
+//    NSLog(@"view %@ will move to window %@", view, window);
+}
+
+- (void)viewDidMoveToWindow:(UIView *)view {
+//    NSLog(@"view %@ did move to window", view);
```
Comment (Alex009) [inline mvvm-livedata/src/iosMain/cinterop/testproj/ViewController.m:19]: uncomment for test project

File: mvvm-livedata/src/iosX64Main/kotlin/LiveDataExtTest.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import dev.icerock.moko.mvvm.livedata.LiveData
+import dev.icerock.moko.mvvm.livedata.cinterop.UIViewLifecycleDelegateProtocol
+import dev.icerock.moko.mvvm.livedata.cinterop.addLifecycleDelegate
+import platform.UIKit.UIView
+import platform.UIKit.UIWindow
+import platform.UIKit.window
+import platform.darwin.NSObject
+import kotlin.native.ref.WeakReference
+
+actual fun <T, V : UIView> LiveData<T>.bind(view: V, setter: V.(T) -> Unit) {
```
Comment (Alex009) [inline mvvm-livedata/src/iosX64Main/kotlin/LiveDataExtTest.kt:16]: change name of file - it's not for test

File: mvvm-state-deprecated/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt [hunk 6]
```diff
@@ -32,29 +32,35 @@ class LiveDataTest {
             }
         }
 
-        assertEquals(actual = dataTransformCounter, expected = 1)
-        assertEquals(actual = mergeWithCounter, expected = 3)
+        // observer required for cold stream
+        mapLd.addObserver { }
+
+        // TODO why expected 2?
```
Comment (Alex009) [inline mvvm-state-deprecated/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt:38]: todo now or later?

File: mvvm-state-deprecated/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt [hunk 7]
```diff
@@ -32,29 +32,35 @@ class LiveDataTest {
...
 
-        assertEquals(actual = dataTransformCounter, expected = 2)
+        assertEquals(actual = dataTransformCounter, expected = 3)
         assertEquals(
             actual = mergeWithCounter,
             expected = 9
-        ) // FIXME: there's an extra mergeWith lambda call
+        )
+        @Suppress("ForbiddenComment")
+        // FIXME: there's an extra mergeWith lambda call
```
Comment (Alex009) [inline mvvm-state-deprecated/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt:63]: create issue for it and add reference

File: mvvm-state/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt [hunk 8]
```diff
@@ -32,29 +32,35 @@ class LiveDataTest {
...
 
-        assertEquals(actual = dataTransformCounter, expected = 2)
+        assertEquals(actual = dataTransformCounter, expected = 3)
         assertEquals(
             actual = mergeWithCounter,
             expected = 9
-        ) // FIXME: there's an extra mergeWith lambda call
+        )
+        @Suppress("ForbiddenComment")
+        // FIXME: there's an extra mergeWith lambda call
```
Comment (Alex009) [inline mvvm-state/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt:63]: create issue for it and add reference

File: mvvm-state/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt [hunk 9]
```diff
@@ -32,29 +32,35 @@ class LiveDataTest {
             }
         }
 
-        assertEquals(actual = dataTransformCounter, expected = 1)
-        assertEquals(actual = mergeWithCounter, expected = 3)
+        // observer required for cold stream
+        mapLd.addObserver { }
+
+        // TODO why expected 2?
```
Comment (Alex009) [inline mvvm-state/src/commonTest/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataTest.kt:38]: todo now or later?

File: sample/ios-app/src/sample5/ValidationAllViewController.swift [hunk 10]
```diff
@@ -6,8 +6,12 @@ import Foundation
 import MultiPlatformLibrary
 import MultiPlatformLibraryMvvm
 
+class MyTextField: UITextField {
+    
+}
```
Comment (Alex009) [inline sample/ios-app/src/sample5/ValidationAllViewController.swift:11]: remove it

File: sample/ios-app/src/sample5/ValidationAllViewController.swift [hunk 11]
```diff
@@ -26,4 +30,32 @@ class ValidationAllViewController: UIViewController {
...
         if(parent == nil) { viewModel.onCleared() }
     }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        DispatchQueue.main.async { [weak self] in
+            self?.navigationController?.popViewController(animated: true)
+        }
+    }
```
Comment (Alex009) [inline sample/ios-app/src/sample5/ValidationAllViewController.swift:40]: remove it

File: sample/ios-app/src/sample5/ValidationAllViewController.swift [hunk 12]
```diff
@@ -26,4 +30,32 @@ class ValidationAllViewController: UIViewController {
...
+            GarbageCollectorKt.collect()
+            GarbageCollectorKt.cycles()
+            return
+        }
+        
+        DispatchQueue.main.async { [weak self] in
+            self?.counter += 1
+            self?.openButton.sendActions(for: .touchUpInside)
+        }
+    }
```
Comment (Alex009) [inline sample/ios-app/src/sample5/ValidationAllViewController.swift:60]: remove it

File: sample/mpp-library/src/commonMain/kotlin/com/icerockdev/library/sample5/ValidationAllViewModel.kt [hunk 13]
```diff
@@ -6,15 +6,35 @@ package com.icerockdev.library.sample5
...
 import dev.icerock.moko.mvvm.livedata.LiveData
 import dev.icerock.moko.mvvm.livedata.MutableLiveData
-import dev.icerock.moko.mvvm.livedata.all
+import dev.icerock.moko.mvvm.livedata.flatMap
 import dev.icerock.moko.mvvm.livedata.map
 import dev.icerock.moko.mvvm.viewmodel.ViewModel
 
+private val myLazyRepository: Repo by lazy { RepoImpl() }
+
+class RepoImpl : Repo {
```
Comment (Alex009) [inline sample/mpp-library/src/commonMain/kotlin/com/icerockdev/library/sample5/ValidationAllViewModel.kt:15]: remove all this changes

**Comment by Alex009** [comment]
https://github.com/icerockdev/moko-mvvm/blob/a20006ed270cfd4c130a161867ab91299294ca61/mvvm-livedata/src/commonMain/kotlin/dev/icerock/moko/mvvm/livedata/LiveDataCoroutines.kt#L25 here should be on active/inactive

**Review by Dorofeev — APPROVED** [review]

**Pull request merged**